data_structures/bloom_filters/bloom_filter/bloom_filter.c

```diff
@@ -0,0 +1,146 @@
+#include "./bloom_filter.h"
+
+#include <math.h>
+
+static inline int murmur_32_scramble(int k)
+{
+    k *= 0xcc9e2d51;
+    k = (k << 15) | (k >> 17);
+    k *= 0x1b873593;
+    return k;
+}
+int murmur3_32(const uint8_t *key, size_t len, int seed)
+{
+    seed = murmur_32_scramble((seed + pow(seed, 3)) * pow(2, seed));
+    int h = seed;
+    int k;
+    /* Read in groups of 4. */
+    for (size_t i = len >> 2; i; i--)
+    {
+        // Here is a source of differing results across endiannesses.
+        // A swap here has no effects on hash properties though.
+        memcpy(&k, key, sizeof(int));
+        key += sizeof(int);
+        h ^= murmur_32_scramble(k);
+        h = (h << 13) | (h >> 19);
+        h = h * 5 + 0xe6546b64;
+    }
+    /* Read the rest. */
+    k = 0;
+    for (unsigned long i = len & 3; i; i--)
+    {
+        k <<= 8;
+        k |= key[i - 1];
+    }
+    // A swap is *not* necessary here because the preceding loop already
+    // places the low bytes in the low places according to whatever endianness
+    // we use. Swaps only apply when the memory is copied in a chunk.
+    h ^= murmur_32_scramble(k);
+    /* Finalize. */
+    h ^= len;
+    h ^= h >> 16;
+    h *= 0x85ebca6b;
+    h ^= h >> 13;
+    h *= 0xc2b2ae35;
+    h ^= h >> 16;
+    return h;
+}
+
+// A union to store the counts as 4-bit buckets instead of 8.
+// Allows for the counting filter to be more space efficient.
+union Bits
+{
+    struct
+    {
+        unsigned int b1 : 1;
+        unsigned int b2 : 1;
+        unsigned int b3 : 1;
+        unsigned int b4 : 1;
+        unsigned int b5 : 1;
+        unsigned int b6 : 1;
+        unsigned int b7 : 1;
+        unsigned int b8 : 1;
+    } nibbles;
```

Comment:
nibbles is unused, but I supposed its existence is necessary to adjust the size of the union to 4 bytes, maybe you should note that in a comment?